### PR TITLE
KEP-6164: Make internal PodSpec type memory-identical to v1

### PR DESCRIPTION
**Pull request by jpbetz**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This has been split out into breakout PRs:

- https://github.com/kubernetes/kubernetes/pull/139851
- https://github.com/kubernetes/kubernetes/pull/139852
- https://github.com/kubernetes/kubernetes/pull/139853
- https://github.com/kubernetes/kubernetes/pull/139854
- https://github.com/kubernetes/kubernetes/pull/139855
- https://github.com/kubernetes/kubernetes/pull/139856
- https://github.com/kubernetes/kubernetes/pull/139857
- https://github.com/kubernetes/kubernetes/pull/139858

See https://github.com/kubernetes/kubernetes/pull/139770 for benchmarks.

#### Does this PR introduce a user-facing change?

```release-note
Pod list operations have been optimized to reduce allocations and memory utilization.
```

### CODE REVIEW

**Comment by serathius** [comment]
Looks awesome 
lgtm

File: pkg/apis/apps/v1/zz_generated.defaults.go [hunk 1]
```diff
@@ -45,6 +45,7 @@ func RegisterDefaults(scheme *runtime.Scheme) error {
 
 func SetObjectDefaults_DaemonSet(in *appsv1.DaemonSet) {
 	SetDefaults_DaemonSet(in)
+	apiscorev1.SetDefaults_PodTemplateSpec(&in.Spec.Template)
```
Comment (liggitt) [inline pkg/apis/apps/v1/zz_generated.defaults.go:48]: (global comment) changes to generated defaults for PodTemplateSpec are surprising...

File: pkg/apis/core/v1/conversion.go [hunk 2]
```diff
@@ -232,90 +230,6 @@ func Convert_v1_ReplicationControllerSpec_To_core_ReplicationControllerSpec(in *
...
-	}
-	// at the this point, autoConvert copied v1.PodIPs -> core.PodIPs
-	// if v1.PodIPs was empty but v1.PodIP is not, then set core.PodIPs[0] with v1.PodIP
-	if len(in.PodIP) > 0 && len(in.PodIPs) == 0 {
-		out.PodIPs = []core.PodIP{
-			{
-				IP: in.PodIP,
-			},
-		}
-	}
```
Comment (liggitt) [inline pkg/apis/core/v1/conversion.go:279]: need to look at the history of why this was placed here, rather than in statusstrategy#PrepareForUpdate or in defaulting ... was this expected to take effect on both read from etcd and status write from older kubelets?

we could handle the static write from older kubelets in statusStrategy#PrepareForUpdate

not sure if we're relying on this to backfill existing pods in etcd on read... if this was needed there, we *could* use Decorator (see Service store's use of `store.Decorator = genericStore.defaultOnRead`)

File: pkg/apis/apps/v1/zz_generated.defaults.go (hunk 1 above)
Comment (liggitt) [reply pkg/apis/apps/v1/zz_generated.defaults.go:48]: mm... I see the serviceaccountname oddness was moved from conversions to defaulting

File: pkg/apis/core/v1/defaults.go [hunk 3]
```diff
@@ -236,6 +236,12 @@ func SetDefaults_PodStatus(obj *v1.PodStatus) {
 	}
 }
 
+func SetDefaults_PodTemplateSpec(obj *v1.PodTemplateSpec) {
+	// drop init container annotations so they don't take effect on legacy kubelets
+	// and don't show up as differences when receiving requests from old clients.
+	obj.Annotations = DropInitContainerAnnotations(obj.Annotations)
```
Comment (jpbetz) [inline pkg/apis/core/v1/defaults.go]: @liggitt I moved this from conversion mutations -> defaulting. After stepping back, this one doesn't sit right to me. It's not a default., but I don't know where else to put it. Could we just drop the requirement that we protect 1.8- kubelets from this? Or do we need to find a way to protect old kubelets from this ~forever?

File: pkg/registry/core/pod/storage/storage.go [hunk 4]
```diff
@@ -250,7 +251,9 @@ func (r *BindingREST) setPodNodeAndMetadata(ctx context.Context, podUID types.UI
 		if pod.Annotations == nil {
 			pod.Annotations = make(map[string]string)
 		}
-		for k, v := range annotations {
+		// Binding annotations bypass Pod defaulting; apply the same legacy
+		// init-container annotation scrub defaulting applies.
+		for k, v := range corev1.DropInitContainerAnnotations(annotations) {
```
Comment (jpbetz) [inline pkg/registry/core/pod/storage/storage.go]: @liggitt Same question here: can we drop this scrub since it only protects 1.8- kubelets?

File: pkg/apis/core/v1/defaults.go (hunk 3 above)
Comment (liggitt) [reply pkg/apis/core/v1/defaults.go]: I'd recommend relocating these cleanups to the write path in a helper we can call from all PrepareForCreate/PrepareForUpdate on pod-spec-containing types.

File: pkg/apis/core/v1/conversion.go (hunk 2 above)
Comment (jpbetz) [reply pkg/apis/core/v1/conversion.go:279]: It was part of the dual stack change in 1.20.  It was being used to backfill on read.

I don't know of any way to *prove* it's safe to drop the backfill on read here.  A pod written to storage in 1.20 or earlier could exist.  I add the decorator.

File: pkg/apis/core/v1/defaults.go (hunk 3 above)
Comment (jpbetz) [reply pkg/apis/core/v1/defaults.go]: Done.

File: pkg/apis/core/v1/conversion.go (hunk 2 above)
Comment (jpbetz) [reply pkg/apis/core/v1/conversion.go:279]: Done.

File: pkg/registry/core/pod/strategy.go [hunk 5]
```diff
@@ -234,15 +234,15 @@ func (podStatusStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.
...
 	}
 
-	syncPodIP(&newPod.Status)
+	SyncPodIP(&newPod.Status)
 	preserveOldObservedGeneration(newPod, oldPod)
 	podutil.DropDisabledPodFields(newPod, oldPod)
 }
 
-// syncPodIP keeps status.podIP and status.podIPs[0] in sync, with podIP
+// SyncPodIP keeps status.podIP and status.podIPs[0] in sync, with podIP
```
Comment (jpbetz) [inline pkg/registry/core/pod/strategy.go]: This needs to be exported since storage depends on strategy and we're sharing this function with storage for the defaulting now.

**Comment by jpbetz** [comment]
@liggitt Feedback applied.  Init container annotation dropping is in `PrepareFor{Create,Update}`, and `PodIP`/`PodIPs` syncing is a storage Decorator.

File: pkg/registry/core/pod/strategy.go
Comment (yedou37) [inline pkg/registry/core/pod/strategy.go:515]: Adding internal `PodStatus.PodIP` seems to require updating `ToSelectableFields()` as well.

This still computes `status.podIP` only from `PodIPs[0]`. Since field selector matching runs before the storage decorator calls `SyncPodIP()`, objects that have `PodIP` set but empty `PodIPs` can be filtered out by `fieldSelector=status.podIP=...`.

**Comment by Jefftree** [comment]
/test pull-kubernetes-gce-master-scale-performance-5000

Numbers look great! would like to see the impact on our 5k scalability test

**Comment by Jefftree** [comment]
Scalability [test results](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/139605/pull-kubernetes-gce-master-scale-performance-5000/2066886558090268672) came back:

- The job is a bit flaky so the failure isn't related
- The internal->v1 conversion is roughly 50% of list allocations, and this PR completely eliminates it
- Overall allocs in scalability tests is still heavily dominated by writes, so there were no notable changes in SLO (this part was a bit surprising, I think if we isolated a benchmark specifically around reads the change would be more apparent)

**Comment by jpbetz** [comment]
@liggitt This is ready for another pass. Thanks @Jefftree and @yedou37 for the scale runs and benchmarks. It appears this is having the desired effect.

File: pkg/apis/core/validation/validation.go [hunk 6]
```diff
@@ -1210,11 +1210,13 @@ func validateProjectionSources(projection *core.ProjectedVolumeSource, projectio
 		}
 		if projPath := srcPath.Child("serviceAccountToken"); source.ServiceAccountToken != nil {
 			numSources++
-			if source.ServiceAccountToken.ExpirationSeconds < 10*60 {
-				allErrs = append(allErrs, field.Invalid(projPath.Child("expirationSeconds"), source.ServiceAccountToken.ExpirationSeconds, "may not specify a duration less than 10 minutes"))
-			}
-			if source.ServiceAccountToken.ExpirationSeconds > 1<<32 {
-				allErrs = append(allErrs, field.Invalid(projPath.Child("expirationSeconds"), source.ServiceAccountToken.ExpirationSeconds, "may not specify a duration larger than 2^32 seconds"))
+			if exp := source.ServiceAccountToken.ExpirationSeconds; exp != nil {
```
Comment (liggitt) [inline pkg/apis/core/validation/validation.go]: if nil, fail as required (and add a unit test for that)

defaulting makes that ~impossible, but this is belt and suspenders

File: pkg/apis/core/validation/validation_test.go
Comment (liggitt) [inline pkg/apis/core/validation/validation_test.go:1]: let's squash this commit with the TokenProjection pointer commit

File: pkg/apis/core/v1/conversion.go [hunk 7]
```diff
@@ -357,17 +347,6 @@ func Convert_v1_PodSpec_To_core_PodSpec(in *v1.PodSpec, out *core.PodSpec, s con
 		out.ServiceAccountName = in.DeprecatedServiceAccount
 	}
 
-	// the host namespace fields have to be handled specially for backward compatibility
-	// with v1.0.0
-	if out.SecurityContext == nil {
-		out.SecurityContext = new(core.PodSecurityContext)
```
Comment (liggitt) [inline pkg/apis/core/v1/conversion.go]: because in the past these fields lived inside spec.securityContext in internal, I *think* roundtripping would result in a non-nil securityContext:

original v1:
```
spec: {
  hostNetwork:true,
  securityContext:null
}
```

converted internal:
```
spec: {
  securityContext:{hostNetwork:true}
}
```

round-tripped v1:
```
spec: {
  hostNetwork: true,
  securityContext: {}
}
```

We can preserve that behavior by defaulting a nil spec.securityContext to {} IF we have true / non-nil host___ fields.

Let's add tests around round-trip behavior of null spec.securityContext for these and preserve externally visible behavior.

File: pkg/apis/core/validation/validation.go [hunk 8]
```diff
@@ -5595,6 +5587,10 @@ func validatePodSpecSecurityContext(securityContext *core.PodSecurityContext, sp
 		}
 	}
 
+	if spec.ShareProcessNamespace != nil && spec.HostPID && *spec.ShareProcessNamespace {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("shareProcessNamespace"), *spec.ShareProcessNamespace, "ShareProcessNamespace and HostPID cannot both be enabled"))
+	}
```
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:5590]: it feels vulnerable to leave validation of fields that don't live inside securityContext inside a function named validatePodSpecSecurityContext, since I could easily see callers skipping calling validatePodSpecSecurityContext if securityContext is nil

let's move this bit out to a validateShareProcessNamespace function and call it alongside validatePodSpecSecurityContext, passing in the pod spec

File: pkg/apis/core/v1/defaults_test.go [hunk 9]
```diff
@@ -1777,6 +1777,38 @@ func roundTrip(t *testing.T, obj runtime.Object) runtime.Object {
...
+	tests := []struct {
+		name                     string
+		serviceAccountName       string
+		deprecatedServiceAccount string
+		expected                 string
+	}{
+		{name: "only serviceAccountName", serviceAccountName: "a", expected: "a"},
+		{name: "only deprecated alias", deprecatedServiceAccount: "b", expected: "b"},
+		{name: "both set, serviceAccountName wins", serviceAccountName: "a", deprecatedServiceAccount: "b", expected: "a"},
+		{name: "neither set", expected: ""},
```
Comment (liggitt) [inline pkg/apis/core/v1/defaults_test.go]: let's be exhaustive for all the behaviors

v1:
```
serviceAccountName="",serviceAccount="" -> ...
serviceAccountName="a",serviceAccount="a" -> ...
serviceAccountName="a",serviceAccount="b" -> ...
serviceAccountName="a",serviceAccount="" -> ...
serviceAccountName="",serviceAccount="a" -> ...
```

if you make the test do `v1 → default → convert to internal → convert to v1 → assert`, you can make sure the test passes against master and this branch identically (even though the behavior moved from conversion to defaulting)

File: pkg/apis/core/types.go [hunk 10]
```diff
@@ -3745,6 +3745,11 @@ type PodSpec struct {
 	// ServiceAccountName is the name of the ServiceAccount to use to run this pod
 	// The pod will be allowed to use secrets referenced by the ServiceAccount
 	ServiceAccountName string
+	// DeprecatedServiceAccount is a deprecated alias for ServiceAccountName.
+	//
+	// Deprecated: Use serviceAccountName instead.
```
Comment (liggitt) [inline pkg/apis/core/types.go:3750]: I don't want *any* internal code (except the defaulting code) using this field ... is this going to flag / prevent *new* use of this field in in-tree code?

File: pkg/registry/core/pod/strategy.go [hunk 11]
```diff
@@ -98,6 +98,10 @@ func (podStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
 	mutatePodAffinity(pod)
 	mutateTopologySpreadConstraints(pod)
 	applyAppArmorVersionSkew(ctx, pod)
+
+	// Admission (e.g. the ServiceAccount plugin) may set ServiceAccountName
+	// after defaulting synced the deprecated alias; re-sync before persisting.
```
Comment (liggitt) [inline pkg/registry/core/pod/strategy.go]: here, for example, I'm surprised linting was happy with setting a deprecated field (this use is fine, but I still hoped linting would flag it)

File: pkg/registry/core/pod/strategy.go [hunk 12]
```diff
@@ -98,6 +98,10 @@ func (podStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
...
+	pod.Spec.DeprecatedServiceAccount = pod.Spec.ServiceAccountName
```
Comment (liggitt) [inline pkg/registry/core/pod/strategy.go]: I don't think we actually need this ... all in-tree setters on internal types (just ServiceAccount admission) set both fields together. All external admission will get re-defaulted, which will sync the fields.

**Comment by liggitt** [comment]
I made it through the `Add DeprecatedServiceAccount to core.PodSpec to match core/v1` commit.

Let's start splitting out logical small changes into their own layered PRs and we can chip away at the changes (keeping this as the final stack is fine)

File: pkg/apis/core/v1/conversion.go (hunk 7 above)
Comment (jpbetz) [reply pkg/apis/core/v1/conversion.go]: Answered on breakout PR: https://github.com/kubernetes/kubernetes/pull/139853/changes/48823180a930c6842db3d4fe61bf800f3ac60c2b#r3439326063

File: pkg/apis/core/validation/validation.go (hunk 6 above)
Comment (jpbetz) [reply pkg/apis/core/validation/validation.go]: Fixed in breakout PR: https://github.com/kubernetes/kubernetes/pull/139851

File: pkg/apis/core/validation/validation.go (hunk 8 above)
Comment (jpbetz) [reply pkg/apis/core/validation/validation.go:5590]: Fixed. As part of the field move I ran into an issue. I left a comment on the breakout PR: https://github.com/kubernetes/kubernetes/pull/139853/changes/48823180a930c6842db3d4fe61bf800f3ac60c2b#r3439355207

File: pkg/apis/core/types.go (hunk 10 above)
Comment (jpbetz) [reply pkg/apis/core/types.go:3750]: I had fixed the lint issues in a separate commit. I've clarified this by including the `//nolint` changes in the corresponding commits in the breakout PRs

File: pkg/registry/core/pod/strategy.go (hunk 12 above)
Comment (jpbetz) [reply pkg/registry/core/pod/strategy.go]: Fixed in breakout

**Comment by jpbetz** [comment]
Dropping this since the breakout PRs replace it.
